### DISCUSSION TITLE
Discussion regarding audio stuffs

### DISCUSSION THREAD

**Discussion by HiFiPhile** [hathach/tinyusb#595] [General]
> What are you going to do with your I2S bridge

@PanRe 

My idea is to replace XMOS ICs, while they work great sometimes it's hard to buy, and I like a more universal solution.

Instead of using MCU's I2S, with a little iCE40 FPGA I can have native DSD output, auto mute on empty FIFO and some crazy (mostly meaningless) stuff like PCM 768K and DSD1024.

On SAM E70 with a 16 bit SRAM interface I got a throughput about 70MB/s :)

If you are interested I can share you more details 😎 

> And why does STM's need an extra physical? Are there not models with HS PHY embedded or what do you need?

They are very limited STM32s with internal HS Phy, they are either too big (144pin LQFP, which occupy more than 3*3cm) or too expensive to make the PCB (most low-cost PCB factories can't do 0.65mm BGA)

**Comment by HiFiPhile** [comment]
One thing I noticed is, I'm not sure it's due to Cortex-M7's pipeline latency, or SAM E70's bus matrix inefficiency, it's very slow to access non-cached memory.

I test the speed by copy data from internal SRAM to external SRAM interface:
If I copy from cached memory I got the maximum throughput (16bit every 2 cycles)
If I copy from non-cached memory between every write I got a gap of more than 10 cycles.

With DMA I can always get the maximum throughput because it do burst access.

As generally USB FIFOs are not cached, I think we have to add DMA support in order to handle high-speed transfer properly.

@hathach Do you have any experience about this behavior on Cortex-M7 ?

**Comment by hathach** [comment]
@HiFiPhile I haven't worked with SAME70 just yet to comment, though on iMXRT 10xx which is also M7, you need to Invalidate mcu cache for peripherals to pick up https://github.com/hathach/tinyusb/blob/master/src/portable/nxp/transdimension/dcd_transdimension.c#L47 . The slow thing probably is due to SAME7x clock, memeory configure or else, I think you will have an better anwser asking in Microchip forum.

**Comment by PanRe** [comment]
> PCM 768K and DSD1024

That's crazy fast :D Yes please share some details, i am always interested in fancy stuff ;) Do you alreaey use your XMOS replacements in an audio project?

**Comment by HiFiPhile** [comment]
I've only done some simple tests the MCU+FPGA combination, including configuring the FPGA by MCU and high-speed communication. 

Currently I'm making a board of SAME70+ICE40HX8K, which can be used as a multi-propose development board or a I2S-bridge.
With a size of 2.5*4.5cm and more than 50 I/Os, I think it'll become a little beast :)

Strangely there are few boards of MCU+FPGA, and most of then can't do high-speed streaming.

![image](https://user-images.githubusercontent.com/4375114/105637955-722f9080-5e70-11eb-8a48-eb756ce5a185.png)

**Comment by HiFiPhile** [comment]
<img src="https://user-images.githubusercontent.com/4375114/105638632-f6cfde00-5e73-11eb-9b98-bb5928787c83.png" width="800">

**Comment by PanRe** [comment]
That looks pretty nice! But you skip the mounting holes? Do you have any case in mind?

**Comment by HiFiPhile** [comment]
There are SMD spacers on the bottom :)

As it's pretty small, I think it's more practical to plug on mother board, for example plug on the DAC board when used as an I2S-Bridge. (Pin headers mounted on the bottom)

![image](https://user-images.githubusercontent.com/4375114/105640742-d6a61c00-5e7f-11eb-94e7-7021cb07ae51.png)

**Reply by PanRe** [reply]
Ah well thats a good idea! How many layers do you use? 4?

**Reply by HiFiPhile** [reply]
Yes 4. At first I was think using 6 layers to fanout all signals, but it turns out 6 layer PCB is still expensive 🤣

**Reply by PanRe** [reply]
I have never used SMD spacers but don't they ned a hole too?
Yeah 6 layers is still expensive ;) Where do you buy it? [Dirty PCB](https://dirtypcbs.com/store/pcbs)?

**Reply by HiFiPhile** [reply]
Some spacers need hole, this one doesn't. Wurth has a complete line of spacers.

https://www.we-online.de/katalog/datasheet/97730606330.pdf

I use JLCPCB, and PCBWay/PCBGOGO for PCBA.

**Comment by HiFiPhile** [comment]
@PanRe I've received 5*PCB half assembled for 35€ :)
![IMG_20210216_161737](https://user-images.githubusercontent.com/4375114/108083029-f4c3fe00-7072-11eb-95a9-815d7083d1cb.jpg)
If you are interested I can "sponsor" you one.

**Comment by PanRe** [comment]
Wow nice! :) Thanks a lot for the offer! Hardware is always welcome! :D But don't you want to test it in advance? ;)--Diese Nachricht wurde von meinem Android Mobiltelefon mit GMX Mail gesendet.Am 16.02.21, 16:22 schrieb HiFiPhile <notifications@github.com>:

   
  @PanRe I've received 5*PCB half assembled for 35€ :)  If you are interested I can "sponsor" you one.

**Reply by HiFiPhile** [reply]
Finally I finished 2 PCBs :) At least USB bootloader works.
I'll test more FPGA stuffs.
![WeChat Image_20210223171211](https://user-images.githubusercontent.com/4375114/108872402-6dd6ce80-75fa-11eb-9443-13a5e0ad305c.png)
![WeChat Photo Editor_20210223171343](https://user-images.githubusercontent.com/4375114/108872483-80e99e80-75fa-11eb-8050-07ada3116848.jpg)

**Comment by PanRe** [comment]
It looks pretty nice! Let me know how it works! :)

**Reply by HiFiPhile** [reply]
Test goes pretty well :) 
Actually I managed to reduce the SRAM interface timing, resulting a throughput of 75/50 MB/s to/from FPGA ! I did a 'memtest' of several hours without I/O error, I think it's stable.

I've just published some documents:
https://github.com/HiFiPhile/IceStrom_Pub

**Comment by PanRe** [comment]
I was wondering how your project is going on? Are you fine with the current audio driver and does all work as you whished for? :)

**Reply by HiFiPhile** [reply]
Honestly I didn't make much progress 😂.... Maybe I need a vacation to regain motivation 😅 I think the audio driver is in good state, I also added fifo support for same70 but not haven't tested yet.  
If you want to play some fpga I can send you one board :)

**Reply by PanRe** [reply]
That is very kind of you but currently i am also lacking time to do anything 😂.

**Comment by HiFiPhile** [comment]
Finally I switched to LPC4322, with the flexibility of SGPIO I no longer need the FPGA.

I made a board with Si5332 as clock generator for MCLK & BCLK, a 74AUP2G79 D Flip-Flop for LRCK & DAT signal shaping.
And a PCM5100 for I2S testing.
![mmexport1627393997056](https://user-images.githubusercontent.com/4375114/127167577-2001e265-5720-43e3-93e6-2eb283a7622f.jpg)

It works pretty well, in theory I can go to 32bit/768kHz but Windows driver only does 32bit/384khz :)

Here is PCM5100 trying to output a square wave of 192kHz :)
![mmexport1627394036629](https://user-images.githubusercontent.com/4375114/127168117-5e2392c4-e502-4a89-b479-c2cfc870d46e.jpg)

**Reply by hathach** [reply]
great to see it works well for you. Out of curiosity, LPC4000 is rather old, NXP have a new LPC55 with highspeed usb and sct peripherals as well. It probably is cheaper than lpc4000, maybe it is a better choice ?

**Reply by HiFiPhile** [reply]
I've also considered LPC55 but it's I2S peripheral is too slow to handle high data rate.

I think the SGPIO of LPC43 it's the only peripheral capable of sampling external clock more than 50MHz.

**Reply by hathach** [reply]
ah right, apparently you have even other peripheral requirement as well. Thanks for sharing.

**Reply by HiFiPhile** [reply]
Finally I've added SPDIF software encoding :)

LPC43xx with SGPIO is a powerhouse, at 24bit/192kHz total CPU usage is about 40%. (A bout 10% without SPDIF)

~No wonder why Segger also use it for J-Link probes~

Video is playback demo with my DAC.

https://user-images.githubusercontent.com/4375114/128610153-55169fd3-6beb-40e1-a589-1771ad922b38.mp4

**Reply by hathach** [reply]
:musical_note: :musical_note: Great to hear you get it working. SGPIO is great, raspberry pico PIO is quite similar (not entirely sure since I don't look deep enough at the two) but got lots of attention from public. Poor NXP, they should do a better marketing/documenting for the peripheral :smile:

**Reply by HiFiPhile** [reply]
Yeah the whole lpc43 line is poorly supported, no LPCxpresso configuration, library is quite old. 
It's very beginner unfriendly, even the SGPIO example on reference manual is wrong 🤣

On i.mx RT there is a shifting register who does similar work like SGPIO but it has slower bus speed (120M vs 204M) despite a much quicker core.

**Comment by PanRe** [comment]
That really sounds quiet nice! Sorry for responding so late, i was moving into a new flat which turned out to be quiet time consuming ^^
I have to admit that SGPIO is something i never encountered before! I have to look into this!

**Reply by HiFiPhile** [reply]
Congrats for your new apartment 🎉

**Reply by hathach** [reply]
@PanRe Congrats to your new :house: :tada: :tada:

**Reply by PanRe** [reply]
Thanks guys 😄

**Reply by hathach** [reply]
I am tired from just barely cleaning my own desk. Moving out, I am not sure how long would it takes, nah never mind 😅😅

**Reply by PanRe** [reply]
It is astonishing how many little crabby stuff you collect over time 😄
